src/OneDriveSdk/Requests/UploadChunkRequest.cs

```diff
@@ -0,0 +1,136 @@
+        /// <param name="client">Client used for sending the chunk.</param>
+        /// <param name="options">Options</param>
+        /// <param name="rangeBegin">Beginning of range of this chunk</param>
+        /// <param name="rangeEnd">End of range of this chunk</param>
+        /// <param name="totalSessionLength">Total session length. This MUST be consistent
+        /// across all chunks.</param>
+        public UploadChunkRequest(
+            string sessionUrl,
+            IBaseClient client,
+            IEnumerable<Option> options,
+            long rangeBegin,
+            long rangeEnd,
+            long totalSessionLength)
+            : base(sessionUrl, client, options)
+        {
+            this.RangeBegin = rangeBegin;
+            this.RangeEnd = rangeEnd;
+            this.TotalSessionLength = totalSessionLength;
+        }
+
+        /// <summary>
+        /// Uploads the chunk using PUT.
+        /// </summary>
+        /// <returns>The status of the upload.</returns>
+        public Task<UploadChunkResult> PutAsync(Stream stream)
+        {
+            return this.PutAsync(stream, CancellationToken.None);
+        }
+
+        /// <summary>
+        /// Uploads the chunk using PUT.
+        /// </summary>
+        /// <param name="stream">Stream of data to be sent in the request. Length must be equal to the length
+        /// of this chunk (as defined by this.RangeLength)</param>
+        /// <param name="cancellationToken">The cancellation token</param>
+        /// <returns>The status of the upload. If UploadSession.AdditionalData.ContainsKey("successResponse")
+        /// is true, then the item has completed, and the value is the created item from the server.</returns>
+        public async Task<UploadChunkResult> PutAsync(Stream stream, CancellationToken cancellationToken)
+        {
+            this.Method = "PUT";
+            using (var response = await this.SendRequestAsync(stream, cancellationToken).ConfigureAwait(false))
+            {
+                if (response.Content != null)
+                {
+                    var responseString = await response.Content.ReadAsStringAsync();
+                    
+                    if (response.StatusCode == HttpStatusCode.Created || response.StatusCode == HttpStatusCode.OK)
+                    {
+                        return new UploadChunkResult
+                            {
+                                ItemResponse =
+                                    this.Client.HttpProvider.Serializer.DeserializeObject<Item>(responseString)
+                            };
+                    }
+                    else
+                    {
+                        return new UploadChunkResult
+                            {
+                                UploadSession =
+                                    this.Client.HttpProvider.Serializer.DeserializeObject<UploadSession>(responseString)
+                            };
```

Comment:
Its possible to get an error instead of a upload session or item response, if you try/catch this block and then thrown an upload error, you can make sure to return a useful error message